### PR TITLE
Sleep for 200µs before each ActivateBestChainStep call

### PR DESCRIPTION
**Pull request by str4d**
This should lower the main thread's likelihood to immediately reacquire `cs_main` after dropping it, which should help `ThreadNotifyWallets` and the RPC methods to acquire `cs_main` more quickly.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -4386,7 +4387,15 @@ bool ActivateBestChain(CValidationState& state, const CChainParams& chainparams,
     CBlockIndex *pindexMostWork = NULL;
     CBlockIndex *pindexNewTip = NULL;
     do {
-        boost::this_thread::interruption_point();
+        // Sleep briefly to allow other threads a chance at grabbing cs_main if
+        // we are connecting a long chain of blocks and would otherwise hold the
+        // lock almost continuously. As of 2023-02-03 the Zcash mainnet chain is
+        // around height 1,972,000; this sleep adds ~6.6 minutes to the time
+        // required to activate the best chain from genesis.
```
[the remark this answers is not in the document]
Comment (daira) [reply src/main.cpp:4394]: Well, there could be additional memory/cache contention between the wallet and other threads, but I would not expect that to have a very large effect.

File: src/main.cpp [hunk 2]
```diff
@@ -4386,7 +4387,15 @@ bool ActivateBestChain(CValidationState& state, const CChainParams& chainparams,
...
+        //
+        // This is defined to be an interruption point.
+        // <https://www.boost.org/doc/libs/1_70_0/doc/html/thread/thread_management.html#interruption_points>
```
Comment (daira) [inline src/main.cpp:4397]:
```suggestion
        // <https://www.boost.org/doc/libs/1_81_0/doc/html/thread/thread_management.html#interruption_points>
```
1.81.0 is the version of boost we [currently depend on](https://github.com/zcash/zcash/blob/feec543abf274d2bb7620720fc8f6833283571b7/depends/packages/boost.mk#L2).

File: src/main.cpp (hunk 1 above)
Comment (daira) [inline src/main.cpp:4394]:
```suggestion
        // around height 1,972,000; this total time slept when activating the
        // best chain from genesis is ~6.6 minutes. In practice this allows the
        // internal wallet, if it is enabled, to keep up with the connected
        // blocks, reducing the overall time until the node becomes usable.
```

**Review by daira — DISMISSED** [review]
utACK, modulo @str4d's testing.

**Comment by str4d** [comment]
I reindexed with e2cd1b761fe556bc6d61849346902c3611530307 and confirmed that it fixes my problem. The graph below shows two reindexes, one without the fix and then one (still ongoing) with the fix.
- The first reindex shows `ThreadNotifyWallets` significantly lagging the main chain until the latter reaches the tip, and then the wallet jumps forward.
- The second reindex shows `ThreadNotifyWallets` basically tracking the main chain height-for-height.

image

**Review by nuttycom — APPROVED** [review]
utACK

**Pull request merged**